cpp/src/gandiva/function_registry.cc

```diff
@@ -41,42 +62,74 @@ FunctionRegistry::iterator FunctionRegistry::back() const {
 
-  auto v1 = GetArithmeticFunctionRegistry();
-  pc_registry_.insert(std::end(pc_registry_), v1.begin(), v1.end());
-  auto v2 = GetDateTimeFunctionRegistry();
-  pc_registry_.insert(std::end(pc_registry_), v2.begin(), v2.end());
+  ARROW_RETURN_IF(!buffer_or_error,
+                  Status::IOError("Could not load module from bitcode file: ",
+                                  bitcode_file_path +
+                                      " Error: " + buffer_or_error.getError().message()));
 
-  auto v3 = GetHashFunctionRegistry();
-  pc_registry_.insert(std::end(pc_registry_), v3.begin(), v3.end());
+  return std::move(buffer_or_error.get());
+}
 
-  auto v4 = GetMathOpsFunctionRegistry();
-  pc_registry_.insert(std::end(pc_registry_), v4.begin(), v4.end());
+Status FunctionRegistry::Register(const std::vector<NativeFunction>& funcs,
+                                  const std::string& bitcode_path) {
+  ARROW_ASSIGN_OR_RAISE(auto llvm_buffer, GetBufferFromFile(bitcode_path));
+  auto buffer = std::make_shared<LLVMMemoryArrowBuffer>(std::move(llvm_buffer));
+  return Register(funcs, std::move(buffer));
+}
 
-  auto v5 = GetStringFunctionRegistry();
-  pc_registry_.insert(std::end(pc_registry_), v5.begin(), v5.end());
+arrow::Status FunctionRegistry::Register(const std::vector<NativeFunction>& funcs,
+                                         std::shared_ptr<arrow::Buffer> bitcode_buffer) {
+  bitcode_memory_buffers_.emplace_back(std::move(bitcode_buffer));
+  for (const auto& func : funcs) {
+    ARROW_RETURN_NOT_OK(FunctionRegistry::Add(func));
+  }
+  return Status::OK();
+}
 
-  auto v6 = GetDateTimeArithmeticFunctionRegistry();
-  pc_registry_.insert(std::end(pc_registry_), v6.begin(), v6.end());
-  for (auto& elem : pc_registry_) {
-    for (auto& func_signature : elem.signatures()) {
-      map.insert(std::make_pair(&(func_signature), &elem));
+const std::vector<std::shared_ptr<arrow::Buffer>>& FunctionRegistry::GetBitcodeBuffers()
+    const {
+  return bitcode_memory_buffers_;
+}
+
+arrow::Result<std::shared_ptr<FunctionRegistry>> MakeDefaultFunctionRegistry() {
+  auto registry = std::make_shared<FunctionRegistry>();
+  for (auto const& funcs :
+       {GetArithmeticFunctionRegistry(), GetDateTimeFunctionRegistry(),
+        GetHashFunctionRegistry(), GetMathOpsFunctionRegistry(),
+        GetStringFunctionRegistry(), GetDateTimeArithmeticFunctionRegistry()}) {
+    for (auto const& func_signature : funcs) {
+      ARROW_RETURN_NOT_OK(registry->Add(func_signature));
     }
   }
-
-  return map;
+  return std::move(registry);
 }
 
-const NativeFunction* FunctionRegistry::LookupSignature(
-    const FunctionSignature& signature) const {
-  auto got = pc_registry_map_.find(&signature);
-  return got == pc_registry_map_.end() ? nullptr : got->second;
+std::shared_ptr<FunctionRegistry> default_function_registry() {
+  static auto maybe_default_registry = MakeDefaultFunctionRegistry();
+  return *maybe_default_registry;
```

Comment:
Can we avoid calling `operator*()` for each `default_function_registry()` call?

```suggestion
  static auto default_registry = *MakeDefaultFunctionRegistry();
  return default_registry;
```

Reply:
Sure. Updated.